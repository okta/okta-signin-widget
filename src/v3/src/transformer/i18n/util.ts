/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
