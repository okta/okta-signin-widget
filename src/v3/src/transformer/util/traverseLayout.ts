/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
