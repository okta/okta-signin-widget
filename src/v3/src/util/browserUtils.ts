/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
