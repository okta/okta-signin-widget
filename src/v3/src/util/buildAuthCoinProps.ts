/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
