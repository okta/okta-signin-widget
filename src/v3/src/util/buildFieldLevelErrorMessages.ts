/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
