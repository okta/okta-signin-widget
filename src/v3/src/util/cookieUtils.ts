/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import Cookies from 'js-cookie';

const DEFAULT_EXPIRY_DAYS = 365;
const DEFAULT_COOKIE_PATH = '/';
const LAST_USERNAME_COOKIE_NAME = 'ln';
const USER_AUTHENTICATED_COOKIE_NAME = 'isuauth';

export const setCookieString = (
  name: string,
  value: string,
  expiryDays?: number,
  path?: string,
): void => {
  Cookies.set(
    name,
    value,
    { expires: expiryDays ?? DEFAULT_EXPIRY_DAYS, path: path ?? DEFAULT_COOKIE_PATH },
  );
};

export const getCookieString = (name: string): string => Cookies.get(name) as string;

export const getCookieUserAuthenticated = (): string => getCookieString(
  USER_AUTHENTICATED_COOKIE_NAME,
);

export const removeCookie = (name: string): void => Cookies.remove(name);

export const setUsernameCookie = (username: string): void => setCookieString(
  LAST_USERNAME_COOKIE_NAME,
  username,
);

export const setCookieUserAuthenticated = (): void => {
  Cookies.set(
    USER_AUTHENTICATED_COOKIE_NAME,
    '1',
    { expires: 1, path: DEFAULT_COOKIE_PATH },
  );
};

export const getUsernameCookie = (): string => getCookieString(LAST_USERNAME_COOKIE_NAME);

export const removeUsernameCookie = (): void => removeCookie(LAST_USERNAME_COOKIE_NAME);

export const removeUserAuthenticatedCookie = (): void => removeCookie(
  USER_AUTHENTICATED_COOKIE_NAME,
);
