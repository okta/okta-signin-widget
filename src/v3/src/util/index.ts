/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
