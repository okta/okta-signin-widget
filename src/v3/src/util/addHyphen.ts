/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
