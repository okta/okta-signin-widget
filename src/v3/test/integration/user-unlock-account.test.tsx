/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { within, waitFor } from '@testing-library/preact';
import { createAuthJsPayloadArgs, setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/unlock-account/default.json';

describe('user-unlock-account', () => {
  it('renders form', async () => {
    const { container, findByText, findByLabelText } = await setup({ mockResponse });
    await findByText(/Unlock account\?/);
    await findByLabelText(/Username/);
    expect(container).toMatchSnapshot();
  });

  it('should display client-side validation errors when trying to submit the flow without a username', async () => {
    const {
      authClient, user, findByTestId, findAllByRole,
    } = await setup({ mockResponse });

    await findByTestId('identifier') as HTMLInputElement;
    const emailAuthenticatorButton = await findByTestId('okta_email');

    await user.click(emailAuthenticatorButton);
    expect(authClient.options.httpRequestClient).not.toHaveBeenCalled();
    const [alertBox] = await findAllByRole('alert');
    within(alertBox).findByText(/We found some errors/);
    const identifierError = await findByTestId('identifier-error');
    expect(identifierError.textContent).toEqual('This field cannot be left blank');
  });

  describe('send corrent payload', () => {
    it('when select email authenticator', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({ mockResponse });
      const titleElement = await findByText(/Unlock account\?/);
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      const emailAuthenticatorButton = await findByTestId('okta_email');
      await user.click(emailAuthenticatorButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/challenge', {
          identifier: 'testuser@okta.com',
          authenticator: {
            id: 'aut2h3fft0FbVvFYV1d7',
          },
        }),
      );
    });

    it('when select phone authenticator', async () => {
      const {
        authClient, user, findByTestId, findByText,
      } = await setup({ mockResponse });

      const titleElement = await findByText(/Unlock account\?/);
      await waitFor(() => expect(titleElement).toHaveFocus());
      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      const phoneAuthenticatorButton = await findByTestId('phone_number');
      await user.click(phoneAuthenticatorButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/challenge', {
          identifier: 'testuser@okta.com',
          authenticator: {
            id: 'aut2h3fft10VeLDPD1d7',
          },
        }),
      );
    });

    it('when click "Back to sign in" button', async () => {
      const { authClient, user, findByTestId } = await setup({
        mockResponse,
        widgetOptions: {
          useInteractionCodeFlow: false,
        },
      });

      const cancelButton = await findByTestId('cancel');
      await user.click(cancelButton);
      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        ...createAuthJsPayloadArgs('POST', 'idp/idx/cancel'),
      );
    });
  });

  describe('useInteractionCodeFlow', () => {
    it('should restart transaction when clicking "Back to sign in" link', async () => {
      const { authClient, user, findByTestId } = await setup({
        mockResponse,
        widgetOptions: {
          useInteractionCodeFlow: true,
        },
      });

      const cancelButton = await findByTestId('cancel');
      await user.click(cancelButton);
      expect(authClient.options.httpRequestClient).toHaveBeenNthCalledWith(1,
        'POST', 'http://localhost:3000/oauth2/default/v1/interact', expect.any(Object));
      expect(authClient.options.httpRequestClient).toHaveBeenNthCalledWith(2,
        ...createAuthJsPayloadArgs(
          'POST', 'idp/idx/introspect', {
            interactionHandle: 'fake-interactionhandle',
          },
          'application/ion+json; okta-version=1.0.0',
          'application/ion+json; okta-version=1.0.0',
        ));
    });
  });
});
