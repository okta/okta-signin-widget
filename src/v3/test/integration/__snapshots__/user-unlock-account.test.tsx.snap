// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user-unlock-account renders form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <main
      class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
      data-commit="b9bbc0140703c3fbf0e2e58920362e70"
      data-version="0.0.0"
      dir="ltr"
      id="okta-sign-in"
      lang="en"
    >
      <div
        class="siwContainer MuiBox-root emotion-2"
      >
        <div
          class="okta-sign-in-header auth-header siwHeader MuiBox-root emotion-3"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 emotion-4"
          />
        </div>
        <div
          class="MuiBox-root emotion-5"
        >
          <form
            class="o-form siwForm"
            data-se="o-form"
            novalidate=""
          >
            <div
              class="MuiBox-root emotion-6"
            >
              <div
                class="MuiBox-root emotion-7"
              >
                <div
                  class="MuiBox-root emotion-8"
                >
                  <h2
                    class="MuiTypography-root MuiTypography-h4 emotion-9"
                    data-se="o-form-head"
                    id="select-authenticator-unlock-account_Title_Unlock_account_1"
                  >
                    Unlock account?
                  </h2>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-7"
              >
                <div
                  class="MuiBox-root emotion-11"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiFormLabel-colorPrimary MuiInputLabel-root emotion-12"
                    for="identifier"
                  >
                    Username
                  </label>
                  <div
                    class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth emotion-13"
                    required="true"
                  >
                    <input
                      aria-describedby="select-authenticator-unlock-account_Title_Unlock_account_1"
                      aria-invalid="false"
                      autocomplete="username"
                      class="MuiInputBase-input emotion-14"
                      data-se="identifier"
                      id="identifier"
                      name="identifier"
                      required=""
                      type="text"
                    />
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-7"
              >
                <ul
                  class="MuiBox-root emotion-16"
                >
                  <li
                    class="MuiBox-root emotion-7"
                  >
                    <button
                      aria-describedby="select-authenticator-unlock-account_Title_Unlock_account_1 okta_email_0-ctaLabel"
                      aria-labelledby="okta_email_0-label"
                      class="authenticator-row authButton MuiBox-root emotion-18"
                      data-se="authenticator-button"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="authenticator-icon-container MuiBox-root emotion-11"
                        data-se="authenticator-icon"
                      >
                        <div
                          class="iconContainer mfa-okta-email authenticator-icon MuiBox-root emotion-20"
                          data-se="factor-beacon"
                        >
                          <svg
                            aria-labelledby="okta_email_0"
                            fill="none"
                            height="48"
                            role="img"
                            viewBox="0 0 48 48"
                            width="48"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <title
                              id="okta_email_0"
                            >
                              Email Authentication
                            </title>
                            <circle
                              class="siwFillBg"
                              cx="24"
                              cy="24"
                              fill="#F5F5F6"
                              r="24"
                            />
                            <path
                              class="siwFillSecondary"
                              d="M32 18v1h5.3L32 24.29v1.42l6-6.01v12.8a1.5 1.5 0 0 1-1.5 1.5h-17a1.5 1.5 0 0 1-1.5-1.5V31h-1v1.5a2.5 2.5 0 0 0 2.5 2.5h17a2.5 2.5 0 0 0 2.5-2.5V18h-7Z"
                              fill="#A7B5EC"
                            />
                            <path
                              class="siwFillPrimary"
                              d="M9 13v14.5a2.5 2.5 0 0 0 2.5 2.5h17a2.5 2.5 0 0 0 2.5-2.5V13H9Zm20.293 1-8.81 8.81L10.794 14h18.499ZM28.5 29h-17a1.5 1.5 0 0 1-1.5-1.5V14.63l10.517 9.56L30 14.707V27.5a1.5 1.5 0 0 1-1.5 1.5Z"
                              fill="#00297A"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="authenticator-description infoSection MuiBox-root emotion-11"
                      >
                        <h3
                          class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-22"
                          data-se="authenticator-button-label"
                          id="okta_email_0-label"
                        >
                          Email
                        </h3>
                        <div
                          class="cta-button authenticator-button actionName MuiBox-root emotion-11"
                          data-se="okta_email"
                        >
                          <span
                            class="button select-factor link-button MuiBox-root emotion-24"
                            data-se="cta-button-label"
                            id="okta_email_0-ctaLabel"
                          >
                            Select
                          </span>
                          <svg
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-25"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M13.2929 7.5L9.14645 3.35355L9.85355 2.64645L14.8536 7.64645C15.0488 7.84171 15.0488 8.15829 14.8536 8.35355L9.85355 13.3536L9.14645 12.6464L13.2929 8.5H1V7.5H13.2929Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              Select
                            </title>
                          </svg>
                        </div>
                      </div>
                    </button>
                  </li>
                  <li
                    class="MuiBox-root emotion-7"
                  >
                    <button
                      aria-describedby="select-authenticator-unlock-account_Title_Unlock_account_1 phone_number_1-ctaLabel"
                      aria-labelledby="phone_number_1-label"
                      class="authenticator-row authButton MuiBox-root emotion-18"
                      data-se="authenticator-button"
                      tabindex="0"
                      type="button"
                    >
                      <div
                        class="authenticator-icon-container MuiBox-root emotion-11"
                        data-se="authenticator-icon"
                      >
                        <div
                          class="iconContainer mfa-okta-phone authenticator-icon MuiBox-root emotion-20"
                          data-se="factor-beacon"
                        >
                          <svg
                            aria-labelledby="phone_number_1"
                            fill="none"
                            height="48"
                            role="img"
                            viewBox="0 0 48 48"
                            width="48"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <title
                              id="phone_number_1"
                            >
                              Voice Call Authentication
                            </title>
                            <circle
                              class="siwFillBg"
                              cx="24"
                              cy="24"
                              fill="#F5F5F6"
                              r="24"
                            />
                            <path
                              class="siwFillPrimary"
                              d="M21.5 37h-9a2.6 2.6 0 0 1-2.5-2.682V14.682A2.6 2.6 0 0 1 12.5 12h9a2.6 2.6 0 0 1 2.5 2.682v19.636A2.6 2.6 0 0 1 21.5 37Zm-9-24a1.6 1.6 0 0 0-1.5 1.682v19.636A1.6 1.6 0 0 0 12.5 36h9a1.6 1.6 0 0 0 1.5-1.682V14.682A1.6 1.6 0 0 0 21.5 13h-9Z"
                              fill="#00297A"
                            />
                            <path
                              class="siwFillPrimary"
                              d="M19 34h-4v1h4v-1Zm-1-19a1 1 0 1 0-2 0 1 1 0 0 0 2 0Z"
                              fill="#00297A"
                            />
                            <path
                              class="siwFillSecondary"
                              d="M34.606 34.606 33.9 33.9a13.998 13.998 0 0 0 0-19.8l.707-.707a15 15 0 0 1 0 21.212l-.001.001Z"
                              fill="#A7B5EC"
                            />
                            <path
                              class="siwFillSecondary"
                              d="m31.071 31.071-.707-.707a9 9 0 0 0 0-12.728l.707-.707a10 10 0 0 1 0 14.142Zm-3.535-3.536-.707-.707a4 4 0 0 0 0-5.656l.707-.707a5 5 0 0 1 0 7.07Z"
                              fill="#A7B5EC"
                            />
                          </svg>
                        </div>
                      </div>
                      <div
                        class="authenticator-description infoSection MuiBox-root emotion-11"
                      >
                        <h3
                          class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-22"
                          data-se="authenticator-button-label"
                          id="phone_number_1-label"
                        >
                          Phone
                        </h3>
                        <div
                          class="cta-button authenticator-button actionName MuiBox-root emotion-11"
                          data-se="phone_number"
                        >
                          <span
                            class="button select-factor link-button MuiBox-root emotion-24"
                            data-se="cta-button-label"
                            id="phone_number_1-ctaLabel"
                          >
                            Select
                          </span>
                          <svg
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-25"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M13.2929 7.5L9.14645 3.35355L9.85355 2.64645L14.8536 7.64645C15.0488 7.84171 15.0488 8.15829 14.8536 8.35355L9.85355 13.3536L9.14645 12.6464L13.2929 8.5H1V7.5H13.2929Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              Select
                            </title>
                          </svg>
                        </div>
                      </div>
                    </button>
                  </li>
                </ul>
              </div>
              <div
                class="MuiBox-root emotion-7"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-36"
                  data-se="cancel"
                  href="javascript:void(0)"
                >
                  Back to sign in
                </a>
              </div>
            </div>
          </form>
        </div>
      </div>
    </main>
  </div>
</div>
`;
